go/vt/vtgate/planbuilder/operators/sharded_routing.go

```diff
@@ -0,0 +1,619 @@
+
+	RouteOpCode engine.Opcode
+
+	// SeenPredicates contains all the predicates that have had a chance to influence routing.
+	// If we need to replan routing, we'll use this list
+	SeenPredicates []sqlparser.Expr
+}
+
+var _ Routing = (*ShardedRouting)(nil)
+
+func newShardedRouting(vtable *vindexes.Table, id semantics.TableSet) Routing {
+	routing := &ShardedRouting{
+		RouteOpCode: engine.Scatter,
+		keyspace:    vtable.Keyspace,
+	}
+
+	if vtable.Pinned != nil {
+		// Pinned tables have their keyspace ids already assigned.
+		// Use the Binary vindex, which is the identity function
+		// for keyspace id.
+		routing.RouteOpCode = engine.EqualUnique
+		vindex, _ := vindexes.NewBinary("binary", nil)
+		routing.Selected = &VindexOption{
+			Ready:       true,
+			Values:      []evalengine.Expr{evalengine.NewLiteralString(vtable.Pinned, collations.TypedCollation{})},
+			ValueExprs:  nil,
+			Predicates:  nil,
+			OpCode:      engine.EqualUnique,
+			FoundVindex: vindex,
+			Cost: Cost{
+				OpCode: engine.EqualUnique,
+			},
+		}
+
+	}
+	for _, columnVindex := range vtable.ColumnVindexes {
+		routing.VindexPreds = append(routing.VindexPreds, &VindexPlusPredicates{ColVindex: columnVindex, TableID: id})
+	}
+	return routing
+}
+
+func (tr *ShardedRouting) isScatter() bool {
+	return tr.RouteOpCode == engine.Scatter
+}
+
+// tryImprove rewrites the predicates for this query to see if we can produce a better plan.
+// The rewrites are two:
+//  1. first we turn the predicate a conjunctive normal form - an AND of ORs.
+//     This can sometimes push a predicate to the top so it's not hiding inside of an OR
+//  2. If that is not enough, an additional rewrite pass is performed where we try to
+//     turn ORs into IN, which is easier for the planner to plan
+func (tr *ShardedRouting) tryImprove(ctx *plancontext.PlanningContext, queryTable *QueryTable) (Routing, error) {
+	oldPredicates := queryTable.Predicates
+	queryTable.Predicates = nil
+	tr.SeenPredicates = nil
+	var routing Routing = tr
+	var err error
+	for _, pred := range oldPredicates {
+		rewritten := sqlparser.RewritePredicate(pred)
+		predicates := sqlparser.SplitAndExpression(nil, rewritten.(sqlparser.Expr))
+		for _, predicate := range predicates {
+			queryTable.Predicates = append(queryTable.Predicates, predicate)
+
+			routing, err = UpdateRoutingLogic(ctx, predicate, routing)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	sr, ok := routing.(*ShardedRouting)
+	if !ok {
+		return routing, nil
+	}
+
+	if sr.isScatter() {
+		// if we _still_ haven't found a better route, we can run this additional rewrite on any ORs we have
+		for _, expr := range queryTable.Predicates {
+			or, ok := expr.(*sqlparser.OrExpr)
+			if !ok {
+				continue
+			}
+			for _, predicate := range sqlparser.ExtractINFromOR(or) {
+				routing, err = UpdateRoutingLogic(ctx, predicate, routing)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+
+	return routing, nil
```

Comment:
nit: unrelated to the refactor. It is easier to follow.

```go
if !sr.isScatter() {
   return routing, nil
}
....
```

Reply:
hard agree